### PR TITLE
[GEN-867] Add validation rule to check if INT_DOD >= INT_CONTACT

### PR DESCRIPTION
**Pull request by danlu1**
Problem:

Lack of validation check on whether INT_DOD >= INT_CONTACT

Solution:

Add the validation check to compare INT_DOD and INT_CONTACT when both of them exist and not NAs.

Testing:
Unit test has been added.

### CODE REVIEW

File: genie_registry/clinical.py [hunk 1]
```diff
@@ -228,6 +228,59 @@ def _check_year_death_validity_message(
...
+    INT_DOD should alway be greater than or equal to INT_CONTACT when they are both available.
+    This function checks if INT_DOD >= INT_CONTACT and returns row indices of invalid INT_DOD rows.
+
+    Args:
+        clinicaldf: Clinical Data Frame
+
+    Returns:
+        pd.Index: The row indices of the row with INT_DOD < INT_CONTACT in the input clinical data
+    """
+    # Generate temp dataframe to handle datatype mismatch in a column
```
Comment (BryanFauble) [inline genie_registry/clinical.py:240]: Try to limit the amount of in-line comments you add. In a lot of cases it has the unintended affect of reducing readability. The code you wrote does well at telling us what the code is doing. That should be the source of truth.

File: genie_registry/clinical.py [hunk 2]
```diff
@@ -228,6 +228,59 @@ def _check_year_death_validity_message(
...
+    temp = clinicaldf[["INT_DOD", "INT_CONTACT"]].copy()
+    # Convert INT_DOD and INT_CONTACT to numeric, coercing errors to NaN
+    temp["INT_DOD"] = pd.to_numeric(temp["INT_DOD"], errors="coerce")
+    temp["INT_CONTACT"] = pd.to_numeric(temp["INT_CONTACT"], errors="coerce")
+    # Compare rows with numeric values in both columns and returns comparion results("True"/"False")
```
Comment (BryanFauble) [inline genie_registry/clinical.py]: This inline comment is misleading. You are only checking for False.

File: genie_registry/clinical.py [hunk 3]
```diff
@@ -228,6 +228,59 @@ def _check_year_death_validity_message(
...
+        (pd.isna(temp["INT_DOD"]) | pd.isna(temp["INT_CONTACT"])),
+        "N/A",
+        temp["INT_DOD"] >= temp["INT_CONTACT"],
+    )
+    invalid_int_dod = temp[temp["check_result"] == "False"]
+    return invalid_int_dod.index
+
+
+def _check_int_dod_validity_message(
+    invalid_int_dod_indices: pd.Index,
```
Comment (BryanFauble) [inline genie_registry/clinical.py:256]: Is the type here a list of pd.Index? The length check below suggests it's an iterator of values.

Comment (BryanFauble) [reply genie_registry/clinical.py:256]: Might also just be the pandas specific way of doing things that is not super intuitive

File: genie_registry/clinical.py [hunk 4]
```diff
@@ -956,6 +1009,16 @@ def _validate(self, clinicaldf):
...
         else:
             total_error.write("Patient Clinical File: Must have DEAD column.\n")
+
+        # CHECK: INT DOD against INT CONTACT
+        has_int_dod_and_contact = process_functions.checkColExist(
+            clinicaldf, ["INT_DOD", "INT_CONTACT"]
+        )
+        if has_int_dod_and_contact:
+            invalid_int_dod_indices = _check_int_dod_validity(clinicaldf)
+            errors, warnings = _check_int_dod_validity_message(invalid_int_dod_indices)
```
Comment (BryanFauble) [inline genie_registry/clinical.py:1015]: Nit:
`errors, _`
As warnings is not used

**Review by BryanFauble — APPROVED** [review]
A few formatting suggestions, otherwise the business logic looks good!

File: genie_registry/clinical.py [hunk 5]
```diff
@@ -228,6 +228,59 @@ def _check_year_death_validity_message(
...
+        input clinical data has
+    """
+    error = ""
+    warning = ""
+    if len(invalid_int_dod_indices) > 0:
+        error = (
+            "Patient Clinical File: Please double check your INT_DOD and INT_CONTACT columns. "
+            "INT_DOD must be >= INT_CONTACT. "
+            f"There are {len(invalid_int_dod_indices)} row(s) with INT_DOD < INT_CONTACT. "
+            f"Row {invalid_int_dod_indices.tolist()} contain invalid values in the INT_DOD field. Please correct.\n"
```
Comment (rxu17) [inline genie_registry/clinical.py]: This part of the message is bit misleading because it implies that the `INT_DOD` field is the issue when it could be `INT_CONTACT`. You could rephrase this part of the error message to be like:
`The row number(s) this occurs in are: {invalid_year_death_indices.tolist()}. Please correct.`

I think this is happening for the `YEAR_DEATH` AND `YEAR_CONTACT` validation rule function too.

**Review by rxu17 — APPROVED** [review]
LGTM! Just one comment

File: genie_registry/clinical.py (hunk 5 above)
Comment (thomasyu888) [reply genie_registry/clinical.py]: ^ Please fix the above in this PR as well.  Thanks!

**Review by thomasyu888 — APPROVED** [review]
🔥 Fantastic work, and great code reviews!  Thanks team - will pre-approve.

File: genie_registry/clinical.py (hunk 1 above)
Comment (danlu1) [reply genie_registry/clinical.py:240]: Sure. Will do.

File: genie_registry/clinical.py (hunk 2 above)
Comment (danlu1) [reply genie_registry/clinical.py]: Hmmm let me rephrase this.

File: genie_registry/clinical.py (hunk 3 above)
Comment (danlu1) [reply genie_registry/clinical.py:256]: The type is `pandas.core.indexes.range.RangeIndex`. We can either use `len` or `size`. 
Screenshot 2024-04-26 at 11 31 37 AM

File: genie_registry/clinical.py (hunk 4 above)
Comment (danlu1) [reply genie_registry/clinical.py:1015]: This is adapted from https://github.com/Sage-Bionetworks/Genie/blob/3f318022e1ef3e5de8c774ce79f7b521ab5653b4/genie_registry/clinical.py#L475-L535 and we might need to output warnings in the future.
